Support conditional exclude for classes

arneee opened the pull request:
Hi,

after further checking #1098, I've noticed that this functionality just isn't there yet. I've created a first draft PR for supporting condtional excludes on class level.

The PR includes the changes in class metadata and SerializationGraphNavigator as well as the annotation driver. 

Still missing is:
* Evaluate condition in DeserializationGraphNavigator
* Other Metadata drivers
* Unit tests

While all existing tests are passing, I was not able to add the corresponding unit tests for this case. I'm also not sure if there are any unintended side effects by not completely excluding the class if there is an \@Exclude annotation like it is done at the moment, but keeping it and do the evaluation during serialization.

Thanks for your help!

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Doc updated   | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1098
| License       | MIT

goetas commented:
Very nice addition! Thanks a lot! 
Will to a proper review in the next days.

arneee commented:
Hi Asmir,

Cool, thanks! 👍 

The PR definitely still needs some work, but since I only used this library yesterday for the first time, I'm not sure how to do the missing parts, especiallcy the unit tests.

Regards,

  Arne

goetas commented on src/Metadata/ClassMetadata.php:
```diff
@@ -126,6 +126,11 @@ class ClassMetadata extends MergeableClassMetadata
     public $xmlDiscriminatorNamespace;
 
+    /**
+     * @var bool
```
This should not be a boolean, probably `string|null|Expression`

spam312sn commented on src/Exclusion/ExpressionLanguageExclusionStrategy.php:
```diff
@@ -56,4 +57,31 @@ public function shouldSkipProperty(PropertyMetadata $property, Context $navigato
+        ];
+        if ($navigatorContext instanceof SerializationContext) {
+            $variables['object'] = $navigatorContext->getObject();
+        } else {
```
I think you can avoid using `else` by adding default value in `$variables` array:

```diff
        $variables = [
            'context' => $navigatorContext,
            'property_metadata' => $class,
+           'object' => null,
        ];

        if ($navigatorContext instanceof SerializationContext) {
            $variables['object'] = $navigatorContext->getObject();
-       } else {
-           $variables['object'] = null;
        }
```

arneee commented:
Hi @goetas ,

i was working on this again and run into a issue. While serialization works fine, there is one problem during deserialization.

```php
/**
 * @Serializer\Exclude(if="object.expired != true")
 */
class PersonAccount {

    /**
     * @Serializer\Type("string")
     */
    public $name;

    /**
     * @Serializer\Type("boolean")
     */
    public $expired;

}
```

During deserialize() we come to the point where the exclude-If expression is being evaluated. Unfortunately the object is null, since the deserialization has not happened yet and therefore the expression fails.

I see two options.

1. Serialize the object first (including all the properties, childs, etc.) and then evaluate the expression
2. Skip the exclude during deserialization 

What is your optionen on this? Thanks!

goetas commented:
This is a known issue, and `@Serializer\Exclude(if="object && object.expired != true")` should do the job. Do you think is enough?

arneee commented:
That would be the easiest solution. For my use case that would be perfectly fine, since I'm interes in serialization only anyway. For the library I don't know, should deserialization offer the same functionality like serialization? If this is ok for you, it's fine for me.

arneee commented:
Ok, hope I got this right and in sync now. 

I've tried to get it working with the solution you have suggested before, but I wasn't able to completely skip the relevant objects during deserialization. If an object was skipped by a class level exclude-if, the array of the parent element still contained a null item.

For now I've removed the Exclude-If on deserialization, it is only works during serialization.

goetas commented on src/Exclusion/ExpressionLanguageExclusionStrategy.php:
```diff
@@ -31,6 +32,33 @@ public function __construct(ExpressionEvaluatorInterface $expressionEvaluator)
+
+        $variables = [
+            'context' => $navigatorContext,
+            'property_metadata' => $class,
```
I believe that should be called class_metadata

arneee replied:
Yepp!

goetas commented on src/Metadata/ClassMetadata.php:
```diff
@@ -126,6 +126,11 @@ class ClassMetadata extends MergeableClassMetadata
+    /**
+     * @var string
+     */
+    public $excludeIf;
```
Can it be also Expression?

arneee replied:
It should... I've taken over this part from the PropertyMetaData. I will also change it in the PropertyMetadata.

arneee commented:
> Can you please add some tests for the Metadata in xml and yaml?

Done (and now the drivers are actually working). 😃

arneee commented:
> Looks good! Can you please fix the code style issue (you can just run vendor/bin/phpcbf), after that I think it is ready to be merged!

It is coming from GraphNavigatorTest, which I didn't change. phpcbf does not find anything on my installation ("No fixable errors were found")

goetas commented:
Thanks a lot for your excellent work! very very nice! :)

arneee commented:
Hi Asmir,
Thanks for the merge, unfortunately I've noticed a mistake on my side. Currently the condition is inverted, means Exclude(if="false") will actually exclude it. This was due to a copy mistake from @Expose in the drivers as well as confusing test. I have simplified the test to make it more clear and fixed the drivers. Do you need a new PR for this?
